### ISSUE TITLE
Backpropagating through membrane potential

### ISSUE THREAD

**Issue by bauerfe** [synsense/sinabs#66]
Currently it is not possible to backpropagate through the recorded membrane potential of a `LIF` or `IAF` layer. This seems to be intentional, as the gradients are being detached in `sinabs/layers/functional/lif.py`, line 84.

I wonder what speaks against letting users backpropagate through the membrane potential?

**Comment by biphasic** [comment]
https://github.com/synsense/sinabs/blob/develop/sinabs/layers/functional/lif.py#L84 is just for recording purposes and detaches the grads for those tensors, other than that backpropagating through vmem is possible, I'm not sure the whole thing would work otherwise? :thinking:

**Comment by bauerfe** [comment]
Yes, backpropagating through `v_mem` (that is, the current state) is possible.

But not detaching the gradients from the recorded v_mem would make it possible to backprop through the whole history of v_mem, for instance if the v_mem over time is part of a loss function, or even for debugging. 

I just think we should leave that decision to the users, whether they want to do that or not. Currently if feels like capabilities are being limited artificially with no obvious reason.

**Comment by biphasic** [comment]
I see! I misunderstood. Yes that's a good idea I agree 

Le 29 novembre 2022 13:17:01 GMT+00:00, Felix Bauer ***@***.***> a écrit :
>Yes, backpropagating through `v_mem` (that is, the current state) is possible.
>
>But not detaching the gradients from the recorded v_mem would make it possible to backprop through the whole history of v_mem, for instance if the v_mem over time is part of a loss function, or even for debugging. 
>
>I just think we should leave that decision to the users, whether they want to do that or not. Currently if feels like capabilities are being limited artificially with no obvious reason.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/synsense/sinabs/issues/66#issuecomment-1330639761
>You are receiving this because you were assigned.
>
>Message ID: ***@***.***>
-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by bauerfe** [comment]
Alright, I will implement this.
